jax/experimental/optix.py

```diff
@@ -324,6 +324,35 @@ def update_fn(updates, state):  # pylint: disable=missing-docstring
   return InitUpdate(init_fn, update_fn)
 
 
+ApplyEvery = collections.namedtuple("ApplyEvery", "count grad_acc")
+
+
+def apply_every(k=1):
+  """accumulate gradients and apply them every k steps.
+
+  Args:
+    k: apply the update every k steps otherwise accumulate the gradients.
+
+  Returns:
+    An (init_fn, update_fn) tuple.
+  """
+
+  def init_fn(params):
+    grad_acc = tree_multimap(jnp.zeros_like, params)
+    return ApplyEvery(count=jnp.zeros([], jnp.int32), grad_acc=grad_acc)
+
+  def update_fn(updates, state):
+
+    reset = state.count % k
+    emit = reset == (k - 1)
+    grad_acc = tree_multimap(
+        lambda g, ga: (reset == 0) * ga + g, updates, state.grad_acc)
```

Comment:
The name `reset` doesn't actually reflect what's in the variable and I think the guard is wrong (afaik `reset == 0` should actually be `reset != 0`).

I'd suggest:

```python
c = state.count % k
acc = c != 0
grad_acc = tree_multimap(lambda g, ga: acc * ga + g, updates, state.grad_acc)
emit = c == (k - 1)
updates = tree_multimap(lambda ga: emit * ga, grad_acc)
```

It would also be great if there was a test for this to avoid regressions..

Reply:
Thanks Tom for spotting that problem. I'd be happy to do some tests but where is optix tested?

Reply:
https://github.com/google/jax/blob/master/tests/optix_test.py